### PR TITLE
APEXMALHAR-1966: Update casandra output opreator

### PR DESCRIPTION
**Pull request by DT-Priyanka**
Changes:
1. Accept user queries (can be insert/update/delete)
2. Use default field info, as per pojo fields if fieldInfos is not set.
3. Add metrics.

### CODE REVIEW

File: contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java [hunk 1]
```diff
@@ -162,6 +163,30 @@ public void activate(Context.OperatorContext context)
...
+  }
+
+  private void populateFieldInfosFromPojo(ColumnDefinitions rsMetaData)
+  {
+    fieldInfos = Lists.newArrayList();
+    Field[] fields = pojoClass.getDeclaredFields();
+    for (int i = 0; i < rsMetaData.size(); i++) {
+      String columnName = rsMetaData.getName(i);
+      String pojoField = getMatchingField(fields, columnName);
+      if (pojoField != null && pojoField.length() != 0) {
```
Comment (bhupeshchawda) [inline contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java:173]: what happens if pojoField is null? Please check if this will cause problems when no matching field is found.

File: contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java [hunk 2]
```diff
@@ -172,25 +197,32 @@ public void deactivate()
...
         values.append(",").append("?");
       }
     }
-    String statement
-            = "INSERT INTO " + store.keyspace + "."
-            + tablename
-            + " (" + queryfields.toString() + ") "
-            + "VALUES (" + values.toString() + ");";
-    LOG.debug("statement is {}", statement);
+    String statement = "INSERT INTO " + store.keyspace + "." + tablename + " (" + queryfields.toString() + ") " + "VALUES (" + values.toString() + ");";
```
Comment (bhupeshchawda) [inline contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java:229]: have a check for null tablename, since `@NotNull` is now removed

File: contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java [hunk 3]
```diff
@@ -260,6 +292,17 @@ protected Statement setStatementParameters(PreparedStatement updateCommand, Obje
...
   }
 
+  @Override
+  public void processTuple(Object tuple)
+  {
+    try {
+      super.processTuple(tuple);
+      recordsProcessed++;
+    } catch (RuntimeException e) {
+      //super.error.emit(tuple);
```
Comment (bhupeshchawda) [inline contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java]: Not emitting on error port?

Comment (bhupeshchawda) [inline contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java]: Also, please log the exception.

File: contrib/src/test/java/com/datatorrent/contrib/cassandra/CassandraOperatorTest.java [hunk 4]
```diff
@@ -269,15 +269,38 @@ public void testCassandraOutputOperator()
     fieldInfos.add(new FieldInfo("set1", "set1", null));
     fieldInfos.add(new FieldInfo("test", "test", null));
 
-    outputOperator.setStore(transactionalStore);
     outputOperator.setFieldInfos(fieldInfos);
+    outputOperator.input.setup(tpc);
```
Comment (bhupeshchawda) [inline contrib/src/test/java/com/datatorrent/contrib/cassandra/CassandraOperatorTest.java]: I think the setup of port is after the setup of operator.

File: contrib/src/test/java/com/datatorrent/contrib/cassandra/CassandraOperatorTest.java [hunk 5]
```diff
@@ -269,15 +269,38 @@ public void testCassandraOutputOperator()
...
+    Assert.assertEquals("rows in db", 3, outputOperator.getNumOfEventsInStore());
+    outputOperator.getEventsInStore();
+  }
 
+  @Test
+  public void testPopulateFieldInfo()
+  {
+    TestOutputOperator outputOperator = setupForOutputOperatorTest();
     outputOperator.input.setup(tpc);
+    outputOperator.setup(context);
```
Comment (bhupeshchawda) [inline contrib/src/test/java/com/datatorrent/contrib/cassandra/CassandraOperatorTest.java:302]: Check order of setup for input port and operator

File: contrib/src/test/java/com/datatorrent/contrib/cassandra/CassandraOperatorTest.java [hunk 6]
```diff
@@ -301,6 +324,59 @@ public void testCassandraOutputOperator()
     outputOperator.getEventsInStore();
   }
 
+  @Test
+  public void testupdateQueryWithParameters()
+  {
+    TestOutputOperator outputOperator = setupForOutputOperatorTest();
+    outputOperator.input.setup(tpc);
```
Comment (bhupeshchawda) [inline contrib/src/test/java/com/datatorrent/contrib/cassandra/CassandraOperatorTest.java]: Setup port after operator

File: contrib/src/test/java/com/datatorrent/contrib/cassandra/CassandraOperatorTest.java [hunk 7]
```diff
@@ -301,6 +324,59 @@ public void testCassandraOutputOperator()
...
+
+    UUID id = UUID.randomUUID();
+    TestPojo testPojo = new TestPojo(id, 20, "Laura", true, 10, 2.0, new HashSet<Integer>(), new ArrayList<Integer>(), null, new Date(System.currentTimeMillis()));
+    // insert record
+    outputOperator.beginWindow(0);
+    outputOperator.input.process(testPojo);
+    outputOperator.endWindow();
+    Assert.assertEquals("rows in db", 1, outputOperator.getNumOfEventsInStore());
+
+    // update record
```
Comment (bhupeshchawda) [inline contrib/src/test/java/com/datatorrent/contrib/cassandra/CassandraOperatorTest.java:341]: Please have this as a separate test. Don't reset the operator..

File: contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java [hunk 8]
```diff
@@ -50,17 +49,21 @@
...
-  @NotNull
   private String tablename;
+  private String query;
 
   protected final transient ArrayList<DataType> columnDataTypes;
   protected final transient ArrayList<Object> getters;
   protected transient Class<?> pojoClass;
 
+  @AutoMetric
+  private long recordsProcessed;
```
Comment (bhupeshchawda) [inline contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java]: Can you rename it to something else? This seems like the total records processed, not records processed successfully.

File: contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java [hunk 9]
```diff
@@ -102,20 +92,31 @@ public CassandraPOJOOutputOperator()
...
+    super.beginWindow(windowId);
+    recordsProcessed = 0;
+    errorRecords = 0;
+  }
+
+  @Override
   public void activate(Context.OperatorContext context)
   {
+    // clear if it had anything
+    columnDataTypes.clear();
```
Comment (bhupeshchawda) [inline contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java]: why is this necessary?

File: contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java [hunk 10]
```diff
@@ -172,25 +197,32 @@ public void deactivate()
   @Override
   protected PreparedStatement getUpdateCommand()
```
Comment (bhupeshchawda) [inline contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java:201]: Please add javadocs stating the behaviour in case query is null and if not null..

File: contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java (hunk 3 above)
Comment (DT-Priyanka) [inline contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java]: In #215 you have added error port to, AbstractTransactionableStoreOutputOperator operator which is parent to this operator. I will update my pull request once I rebase with latest changes.

File: contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java (hunk 8 above)
Comment (DT-Priyanka) [inline contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java]: done.

File: contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java (hunk 9 above)
Comment (DT-Priyanka) [inline contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java]: It was required as in same test I am first running insert query and then update, let me see if I can refactor that, anyone in product we won't support if a query and fieldInfo changes on fly.

File: contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java (hunk 1 above)
Comment (DT-Priyanka) [inline contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java:173]: That field will be ignored (i.e. won't be added to DB). I will add a error message for same, exception doesn't look right here, whats your opinion?

File: contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java (hunk 3 above)
Comment (DT-Priyanka) [inline contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOOutputOperator.java]: copying relevant code from #215 to remove dependency.

**Comment by DT-Priyanka** [comment]
will rebase and test it again as long time has passed from time pull request is opened.
